openhtf/conf.py

```diff
@@ -43,250 +128,224 @@
 
-class _DeclaredKeys(object):
-  """An object which manages config declarations.
+  def __init__(self, flags, logger, lock, _functools, **kwargs):
+    """Initializes the configuration state.
 
-  This object is a helper for Config.  It provides locked access to a map of
-  declarations and processes configuration values against the declaration.  It
-  must be guarded since declarations are updated at import time and if something
-  is lazily imported we could race between updating the declarations map and
-  reading it to check a config value.
-
-  Not thread-safe, requires an external lock!
-  """
-
-  def __init__(self):
-    self._declared = {}
-
-  def Declare(self, name, declaration):
-    """Adds a declared key to this list of Declared Keys.
-
-    Args:
-      name: The name of this value.
-      declaration: A _DeclaredKeys.DECLARATION object.
-
-    Raises:
-      ConfigurationAlreadyDeclared: If a declaration already exists for
-          this key.
-    """
-    if name in self._declared:
-      raise ConfigurationAlreadyDeclared(name)
-    self._declared[name] = declaration
-
-  def CheckValueAgainstDeclaration(self, name, value):
-    """Checks that the provided value is valid given its declaration.
+    We have to pull everything we need from global scope into here because we
+    will be swapping out the module with this instance and will lose any global
+    references.
 
     Args:
-      name: Name of configuration key.
-      value: Value from configuration.
-
-    Returns:
-      The config value if provided, or the default_value if given.
-
-    Raises:
-      UndeclaredKeyAccessError: If key 'name' is undeclared.
-      MissingRequiredConfigurationKeyError: If key is required and value
-          is None
+      flags: gflags.FLAGS object used to access flag values.
+      logger: Logger to use for logging messages within this class.
+      lock: Threading.Lock to use for locking access to config values.
+      _functools: Reference to the functools module so we can use it internally
+          for decorating methods.
+      **kwargs: Modules we need to access within this class.
     """
-    declaration = self._declared.get(name, None)
-
-    if not declaration:
-      raise UndeclaredKeyAccessError(name)
-
-    if (value is None
-        and declaration.default_value is None
-        and not declaration.optional):
-      raise MissingRequiredConfigurationKeyError(
-          declaration.name, declaration.description)
+    self._flags = flags
+    self._logger = logger
+    self._lock = lock
+    self._functools = _functools
+    self._modules = kwargs
+    self._declarations = {}
+    self._loaded_values = {}
+    self._flag_values = {}
+    for keyval in self._flags['config-value'].value:
+      self._flag_values.setdefault(*keyval.split('=', 1))
+
+    # Everywhere that uses configuration uses this, so we just declare it here.
+    self.Declare('station_id', 'The name of this tester')
```

Comment:
That's actually what I do in some cases. I shove the hostname in for tests that aren't designated a particular station. I think taking it as an actual argument is one possibility, but it's in line with the philosophy of configs being station-specific. As an argument makes getting it from configs difficult, so I'd say it could be both: A default config key specified in `__init__.py` as well as taken by `Test.__init__()` (as `conf.InjectPositionalArgs`? That might reverse the override semantics?).